src/obstacle_detector.cpp

```diff
@@ -307,31 +307,42 @@ float ObstacleDetector::obstacle_angle(bool left)
         float theta = std::atan2(y, x);
         float r_squared = x*x + y*y;
         if (r_squared < back_diag) {
-
-           // left side: y = W, -B < x < F
-           // right side: y = -W, -B < x < F
-           float xi = std::sqrt(r_squared - W*W);
-           if (-B < xi && xi < F) {
-               check_angle(theta, xi, W, left, min_angle);
-               check_angle(theta, xi, -W, left, min_angle);
-           }
-           if (-B < -xi && -xi < F) {
-               check_angle(theta, -xi, W, left, min_angle);
-               check_angle(theta, -xi, -W, left, min_angle);
+           // left line segment: y = W, -B < x < F
+           // right line segment: y = -W, -B < x < F
+           if (W*W < r_squared) {
+               float xi = std::sqrt(r_squared - W*W);
+               if (-B < xi && xi < F) {
+                   if (y > 0) {
+                       check_angle(theta, xi, W, left, min_angle);
+                   }
+                   else {
+                       check_angle(theta, xi, -W, left, min_angle);
+                   }
+               }
+               if (-B < -xi && -xi < F) {
+                   if (y > 0) {
```

Comment:
Why not just std:abs()?

Reply:
The sign of `W` changes depending on the sign of `y`.  This is to check for intersections with the point `(x,y)` on the appropriate side of robot.